Source/AssetRipper.Processing/AnimationClips/Editor/StreamedCurveKey.cs

```diff
@@ -12,96 +10,91 @@ public StreamedCurveKey(int index, Vector3 coefficient, float value)
-			StreamedCurveKey curveKey = new StreamedCurveKey(curveKeyIndex, curveKeyCoef, curveKeyValue);
-			return curveKey;
+			Value = LSV = value;
 		}
 
 		public void Read(ref EndianSpanReader reader)
 		{
 			Index = reader.ReadInt32();
-			Coefficient = new Vector3(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
+			Coefficient = new(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
 			Value = reader.ReadSingle();
+			LSV = Value;
 		}
 
-		/// <summary>
-		/// Calculate value between two KeyframeTpl Float at given time
-		/// </summary>
-		/// <remarks>
-		/// This calculates a unit interval cubic Hermite spline.<br />
-		/// <see href="https://en.wikipedia.org/wiki/Cubic_Hermite_spline"/><br />
-		/// <see href="https://en.wikipedia.org/wiki/Hermite_interpolation"/>
-		/// </remarks>
-		/// <param name="deltaTimeFraction">(time - leftTime) / (rightTime - leftTime)</param>
-		/// <param name="leftValue">lhs.Value</param>
-		/// <param name="leftTangent">lhs.OutSlope * (rightTime - leftTime)</param>
-		/// <param name="rightValue">rhs.Value</param>
-		/// <param name="rightTangent">rhs.OutSlope * (rightTime - leftTime)</param>
-		/// <returns>Value between two keyframes</returns>
-		public static float HermiteInterpolate(float deltaTimeFraction, float leftValue, float leftTangent, float rightValue, float rightTangent)
+		public void CalculateSlopes(float time, float nextTime, StreamedCurveKey nextKey)
 		{
-			float tt = deltaTimeFraction * deltaTimeFraction;
-			float ttt = tt * deltaTimeFraction;
-			float tttx2 = ttt * 2.0f;
-			float ttx3 = tt * 3.0f;
-			float v1 = (deltaTimeFraction + ttt - 2.0f * tt) * leftTangent + (tttx2 - ttx3 + 1.0f) * leftValue;
-			float v2 = (ttt - tt) * rightTangent;
-			float v3 = ttx3 - tttx2;
-			return v1 + v2 + v3 * rightValue;
-		}
-
-		public float CalculateNextInSlope(float deltaTime, float nextValue)
-		{
-			if (deltaTime >= 3.40282347e38f)
+			if (Coefficient != default) // != (0,0,0) => regular slopes
 			{
-				return 0;
+				outSlope = Coefficient.Z;
+				float deltaX = nextTime - time;
+				float _CoeffX = Coefficient.X * deltaX * deltaX;
+				float _CoeffY = Coefficient.Y * deltaX;
+				nextKey.inSlope = 3 * _CoeffX + 2 * _CoeffY + Coefficient.Z;
+				//nextKey.inSlope = float.Round(nextKey.inSlope, 4); // amount of decimal places may be more dynamic
+				if (nextKey.Coefficient == default) // calculate LSV only if next slope calculation needs it
+				{
+					nextKey.LSV = (_CoeffX + _CoeffY + Coefficient.Z) * deltaX + Value;
+				}
+				return;
 			}
-			if (deltaTime < 0.00009999999747378752)
+			if (Value != LSV)
 			{
-				deltaTime = 0.000099999997f;
+				// check percentage difference
+				float diff = Value > LSV ? Value - LSV : LSV - Value;
+				float AbsVal = Value < 0 ? -Value : Value;
+				float AbsLSV = LSV < 0 ? -LSV : LSV;
+				float div = AbsVal > AbsLSV ? AbsVal : AbsLSV;
+				float ROUNDING_ERROR = 1e-5f;
```

Comment:
This seems arbitrary.

Reply:
- The way I calculated `float div` for checking percentage difference was so `div` would never be 0 (greater of 2 positive values), and also `diff/div` would be positive.
- `float ROUNDING_ERROR = 1e-5f` is in fact arbitrary, it should be as small as possible but still being a good discriminator for float rouding errors. With its current value you would interpret the code as "if the percentage diff between left and right side limits is greater than 0.001%, then left and right side values are indeed different/not a rounding error".

Reply:
Make it `const` and add a comment indicating that it's arbitrary.

Reply:
implemented